### PR TITLE
MM-33723: Fix pasting code snippet when text is selected in create message box

### PR DESCRIPTION
**Pull request by sinansonmez**
#### Summary
This PR fixes an issue when there is already a selected text in the create message box and GitHub code snippet is pasted, selection is ignored and current text is duplicated.  

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/19907
JIRA: https://mattermost.atlassian.net/browse/MM-33723

#### Screenshots
**before**

https://user-images.githubusercontent.com/37421564/171742236-aabecbfc-3c5f-4e31-85ed-3610b0f19a15.mov

**after**

https://user-images.githubusercontent.com/37421564/171742345-49bfedc4-a269-408a-a471-25e847495eb2.mov

#### Release Note
```release-note
Bug fixes related to pasting GitHub code snippet when text is selected in create message box
```

### CODE REVIEW

**Comment by sinansonmez** [comment]
The reasons for draft status are explained [here](https://github.com/mattermost/mattermost-server/issues/19907#issuecomment-1145375407)

**Comment by koox00** [comment]
Hi @sinansonmez, thanks for the PR!
It looks like to me that `splitMessageBasedOnCaretPosition` could be aware of selection start and end instead of just caret position, and that might solve the issue we are having. 
Since, AFAICT when we select some text and paste, the `firstPiece` is OK, but the `lastPiece` seems to be including the selection.
Does it make sense to you?

**Comment by sinansonmez** [comment]
Thanks for the suggestion @koox00 When I check `splitMessageBasedOnCaretPosition`, the function only receives `caretPosition` and `message`. 

The `lastPiece` is including the selection. However since `splitMessageBasedOnCaretPosition` doesn't have complete selection info (i.e. selectionStart and selectionEnd), imo it is not possible to use `splitMessageBasedOnCaretPosition` to fix this issue. 

Please let me know if I am missing sth here.

**Comment by koox00** [comment]
> Thanks for the suggestion @koox00 When I check splitMessageBasedOnCaretPosition, the function only receives caretPosition and message.
>
> The lastPiece is including the selection. However since splitMessageBasedOnCaretPosition doesn't have complete selection info (i.e. selectionStart and selectionEnd), imo it is not possible to use splitMessageBasedOnCaretPosition to fix this issue.
>
> Please let me know if I am missing sth here.

@sinansonmez is it possible to make `splitMessageBasedOnCaretPosition` aware of selection end?
If it's not, could we maybe add another function e.g. `splitMessageBasedOnSelection`,
and call that either one conditionally?

**Comment by koox00** [comment]
Furthermore, we should apply changes to both `create_post` (Center channel) and `create_comment` (RHS)

**Comment by sinansonmez** [comment]
@koox00 yes that makes sense. 

I can pass `selectionStart` and `selectionEnd` to `splitMessageBasedOnCaretPosition` then update this function to return correct `firstPiece` and `lastPiece` based on selection

**Comment by sinansonmez** [comment]
> Furthermore, we should apply changes to both `create_post` (Center channel) and `create_comment` (RHS)

thanks @koox00 Actually, I will look at these components: `CreatComment`, `EditPost`, `AdvancedCreatePost` and `AdvancedCreateComment`.

**Comment by sinansonmez** [comment]
@koox00 I ended up creating seperate function to split the message similar to `splitMessageBasedOnCaretPosition ` 

I completed the most part of this PR: Missing part is related to `EditPost` and update tests. 
However in `EditPost`, I found an existing bug which I will try to tackle in this PR. In order to recreate the issue:
1. Edit an existing message
2. Paste a Github code

**Comment by sinansonmez** [comment]
@koox00 Bug is fixed. PR is ready for review

**Comment by hmhealey** [comment]
@koox00 Will you have time to review this, or should we re-assign? I also added a second dev reviewer since this PR was missing one

File: utils/post_utils.test.jsx [hunk 1]
```diff
@@ -971,6 +971,20 @@ describe('PostUtils.splitMessageBasedOnCaretPosition', () => {
...
 });
 
+describe('PostUtils.splitMessageBasedOnTextSelection', () => {
+    const state = {
+        selectionStart: 5,
+        selectionEnd: 12,
+    };
+
+    const message = 'Test Replace Message';
+    it('should return an object with two strings when given context and message', () => {
```
Comment (koox00) [inline utils/post_utils.test.jsx:981]: You could modify a bit this test to add a couple of more test cases with [`test.each`](https://jestjs.io/docs/api#testeachtablename-fn-timeout), just for completion.

**Review by koox00 — APPROVED** [review]
Thanks again @sinansonmez! I think the issue is completely solved now 💪 
I've posted a suggestion in a test, which you can ignore if you like

File: utils/post_utils.test.jsx (hunk 1 above)
Comment (sinansonmez) [reply utils/post_utils.test.jsx]: Good point. I improved the test with `test.each`

**Pull request merged**